### PR TITLE
Switch to CPM

### PR DESCRIPTION
**Pull request by zfergus**
Switch from FetchContent to [CPM](https://github.com/cpm-cmake/CPM.cmake).

* Update Catch2 to v3

### CODE REVIEW

File: cmake/recipes/polysolve.cmake [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+# License: MIT
+
+if(TARGET polysolve)
+    return()
+endif()
+
+message(STATUS "Third-party: creating target 'polysolve'")
+
+# TODO: this requires a conflicting version of Eigen. Reenable when Eigen 3.4+ is available.
+set(POLYSOLVE_WITH_ACCELERATE OFF CACHE BOOL "Enable Apple Accelerate" FORCE)
```
Comment (zfergus) [inline cmake/recipes/polysolve.cmake:11]: There is an issue with conflicting versions of Eigen when this option is enabled.

File: CMakeLists.txt [hunk 2]
```diff
@@ -24,137 +24,98 @@ if(EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/PolyFEMOptions.cmake)
     include(${CMAKE_CURRENT_SOURCE_DIR}/PolyFEMOptions.cmake)
 endif()
 
+# Enable ccache if available
+find_program(CCACHE_PROGRAM ccache)
+if(CCACHE_PROGRAM)
+  option(POLYFEM_WITH_CCACHE "Enable ccache when building PolyFEM" ${IPC_TOOLKIT_TOPLEVEL_PROJECT})
```
Comment (teseoch) [inline CMakeLists.txt:30]: polyfem top, right?

File: CMakeLists.txt [hunk 3]
```diff
@@ -195,66 +186,66 @@ if(POLYFEM_THREADING STREQUAL "CPP")
...
 
 # MSHIO
 include(mshio)
-polyfem_target_link_system_libraries(polyfem PUBLIC mshio)
+target_link_libraries(polyfem PUBLIC mshio)
 
 # Geogram
 include(geogram)
-polyfem_target_link_system_libraries(polyfem PUBLIC geogram)
+target_link_libraries(polyfem PUBLIC geogram)
```
Comment (teseoch) [inline CMakeLists.txt:212]: why remove the other target that suppressed warnings?

File: cmake/recipes/polysolve.cmake (hunk 1 above)
Comment (teseoch) [reply cmake/recipes/polysolve.cmake:11]: Maybe @jdumas knows how?

File: CMakeLists.txt (hunk 2 above)
Comment (zfergus) [reply CMakeLists.txt:30]: oops

File: CMakeLists.txt (hunk 3 above)
Comment (zfergus) [reply CMakeLists.txt:212]: CPM does it for us automatically!

File: cmake/recipes/polysolve.cmake (hunk 1 above)
Comment (jdumas) [reply cmake/recipes/polysolve.cmake:11]: Why not let polysolve pull Eigen directly?

Comment (zfergus) [reply cmake/recipes/polysolve.cmake:11]: I get this compilation error:
```
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/friction/constraints/friction_constraint.cpp:1:
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/friction/constraints/friction_constraint.hpp:3:
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/collisions/collision_constraints.hpp:3:
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/collision_mesh.hpp:5:
In file included from /Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/Core:329:
/Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/src/Core/IndexedView.h:148:14: error: data member instantiated with function type 'bool ()'
  ColIndices m_colIndices;
             ^
/Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/friction/constraints/friction_constraint.cpp:284:25: note: in instantiation of template class 'Eigen::IndexedView<const Eigen::Matrix<double, -1, -1, 0, 6, 12>, Eigen::ArithmeticSequence<long, long, long>, bool ()>' requested here
                        jac_Gamma_wrt_beta(
                        ^
/Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/friction/constraints/friction_constraint.cpp:285:63: error: no member named 'all' in namespace 'Eigen'; did you mean 'Eigen::indexing::all'?
                            Eigen::seqN(i, beta.size(), dim), Eigen::all)
                                                              ^~~~~~~~~~
                                                              Eigen::indexing::all
/Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/src/Core/ArithmeticSequence.h:250:30: note: 'Eigen::indexing::all' declared here
  using Eigen::placeholders::all;
                             ^
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/friction/constraints/friction_constraint.cpp:1:
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/friction/constraints/friction_constraint.hpp:3:
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/collisions/collision_constraints.hpp:3:
In file included from /Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/collision_mesh.hpp:5:
In file included from /Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/Core:290:
In file included from /Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/src/Core/DenseBase.h:662:
/Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/src/Core/util/../../plugins/IndexedViewMethods.h:63:73: error: no matching member function for call to 'ivcCol'
    return ConstReturnType(derived, derived.ivcRow(rowIndices), derived.ivcCol(colIndices));
                                                                ~~~~~~~~^~~~~~
/Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/src/Core/util/../../plugins/IndexedViewMethods.h:239:55: note: in instantiation of member function 'Eigen::DenseBase<Eigen::Matrix<double, -1, -1, 0, 6, 12>>::IndexedViewSelector<Eigen::ArithmeticSequence<long, long, long>, bool (), false, false, true>::run' requested here
  return IndexedViewSelector<RowIndices, ColIndices>::run(derived(), rowIndices, colIndices);
                                                      ^
/Users/zachary/.cache/CPM/ipc-toolkit/ff96ec46f901ecdd6203d75076174d996a5eae0d/src/ipc/friction/constraints/friction_constraint.cpp:284:25: note: in instantiation of function template specialization 'Eigen::DenseBase<Eigen::Matrix<double, -1, -1, 0, 6, 12>>::operator()<Eigen::ArithmeticSequence<long, long, long>, bool (), true>' requested here
                        jac_Gamma_wrt_beta(
                        ^
/Users/zachary/.cache/CPM/eigen/2e68522bbe453e136c9b281a06f112155cbeafd8/Eigen/src/Core/util/../../plugins/IndexedViewMethods.h:34:28: note: candidate template ignored: substitution failure [with Indices = bool ()]: function cannot return function type 'IvcColType<bool ()>' (aka 'bool ()')
inline IvcColType<Indices> ivcCol(const Indices& indices) const {
       ~~~~~~~~~~          ^
3 errors generated.
make[2]: *** [_deps/ipc-toolkit-build/CMakeFiles/ipc_toolkit.dir/src/ipc/friction/constraints/friction_constraint.cpp.o] Error 1
make[1]: *** [_deps/ipc-toolkit-build/CMakeFiles/ipc_toolkit.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Comment (zfergus) [reply cmake/recipes/polysolve.cmake:11]: Created an Issue #240 and merging without the fix for now.

**Pull request merged**